### PR TITLE
Generate TChannel client

### PR DESCRIPTION
**Pull request by ChuntaoLu**

### CODE REVIEW

File: examples/example-gateway/endpoints/baz/baz_simpleservice_method_call_test.go [hunk 1]
```diff
@@ -39,10 +39,11 @@ import (
 var testCallCounter int
 
 func call(
-	ctx context.Context, reqHeaders map[string]string, r *baz.BazRequest,
+	ctx context.Context, reqHeaders map[string]string, args *baz.SimpleService_Call_Args,
```
Comment (zhenghuiwang) [inline examples/example-gateway/endpoints/baz/baz_simpleservice_method_call_test.go]: `SimpleService_Call_Args` is a mix of cap and underscore.

File: codegen/module_system.go [hunk 2]
```diff
@@ -157,10 +167,84 @@ func (generator *HTTPClientGenerator) Generate(
...
+		instance.Directory,
+	)
+
+	clientFilePath, err := filepath.Rel(clientDirectory, clientSpec.GoFileName)
+	if err != nil {
+		clientFilePath = clientSpec.GoFileName
+	}
+
+	// Return the client files
+	files := map[string][]byte{}
```
Comment (zhenghuiwang) [inline codegen/module_system.go]: three lines can be merged into one:

```
return map[string][]byte {
   clientFilePath: client
}, nil
```

File: codegen/method.go [hunk 3]
```diff
@@ -63,6 +63,8 @@ type MethodSpec struct {
 	RequestStruct []StructSpec
 	// Thrift service name the method belongs to.
 	ThriftService string
+	// The thriftrw-generated go package name
+	GenCodePkgName string
 	// Whether the method needs annotation or not.
 	WantAnnot bool
```
Comment (zhenghuiwang) [inline codegen/method.go]: Nit: is `IsAnnotated` a better name than `wantAnnot`?

Comment (ChuntaoLu) [inline codegen/method.go]: The Method may or may not **need** the thrift file to have annotations to be created, hence the verb

File: examples/example-gateway/endpoints/baz/baz_simpleservice_method_call_test.go (hunk 1 above)
Comment (ChuntaoLu) [inline examples/example-gateway/endpoints/baz/baz_simpleservice_method_call_test.go]: It's auto generated by thriftrw, I guess the reason is to be able to tell the service, method from the name

File: codegen/module_system.go (hunk 2 above)
Comment (ChuntaoLu) [inline codegen/module_system.go]: sure, will update in https://github.com/uber/zanzibar/pull/56

File: examples/example-gateway/build/clients/baz/baz.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+		error,
+	)
+}
+
+// NewClient returns a new TChannel client for service baz.
+func NewClient(gateway *zanzibar.Gateway) *BazClient {
+	serviceName := gateway.Config.MustGetString("clients.baz.serviceName")
+	sc := gateway.Channel.GetSubChannel(serviceName)
+
+	ip := gateway.Config.MustGetString("clients.baz.ip")
```
Comment (Raynos) [inline examples/example-gateway/build/clients/baz/baz.go]: In the future we want to support multiple host ports.

File: examples/example-gateway/build/clients/baz/baz.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	var result clientsBazBaz.SimpleService_Call_Result
+
+	success, respHeaders, err := c.client.Call(
+		ctx, c.thriftService, "Call", reqHeaders, args, &result,
+	)
+
+	if err == nil && !success {
+		switch {
+		default:
+			err = errors.New("received no result or unknown exception for Call")
```
Comment (Raynos) [inline examples/example-gateway/build/clients/baz/baz.go]: we should put the client name in this err.

File: codegen/templates/tchannel_client.tmpl [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+)
+
+
+// Interfaces for the service and client for the services defined in the IDL.
+{{/* TODO: case */}}
+{{$clientID := .ClientID -}}
+{{$clientName := title .ClientID | printf "%sClient" -}}
+{{range $svc := .Services}}
+// TChan{{title $clientID}} is the interface that defines the server handler and client interface.
+type TChan{{title $clientID}} interface {
```
Comment (Raynos) [inline codegen/templates/tchannel_client.tmpl]: Do we need this interface ? who uses it ?

File: codegen/templates/tchannel_client.tmpl [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+// {{$clientName}} is the TChannel client for downstream service.
+type {{$clientName}} struct {
+	thriftService string
+	client        zanzibar.TChanClient
+}
+
+{{range .Methods}}
+    {{- if and (eq .RequestType "") (eq .ResponseType "")}}
```
Comment (Raynos) [inline codegen/templates/tchannel_client.tmpl]: This duplication is too heavy, take a look at http_client.tmpl it does not duplicate the entire implementation 4 times.

File: codegen/module_system.go [hunk 8]
```diff
@@ -58,6 +58,16 @@ func NewDefaultModuleSystem(h *PackageHelper) (*module.System, error) {
 		)
 	}
 
+	if err := system.RegisterClassType("client", "tchannel", &TCahnnelClientGenerator{
```
Comment (ChuntaoLu) [inline codegen/module_system.go]: fix in https://github.com/uber/zanzibar/pull/57/commits/bb184bdb681025e1c171837b840b81c89ad5f9f8

File: examples/example-gateway/build/clients/baz/baz.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+// Interfaces for the service and client for the services defined in the IDL.
+
+// TChanBaz is the interface that defines the server handler and client interface.
+type TChanBaz interface {
+	Call(
+		ctx context.Context,
+		reqHeaders map[string]string,
+		args *clientsBazBaz.SimpleService_Call_Args,
+	) (
```
Comment (ChuntaoLu) [inline examples/example-gateway/build/clients/baz/baz.go]: fix in https://github.com/uber/zanzibar/pull/57/commits/7d75a0d649c32002fbdba3b6acf2b53b74fb8dfa

File: examples/example-gateway/build/clients/baz/baz.go (hunk 4 above)
Comment (ChuntaoLu) [inline examples/example-gateway/build/clients/baz/baz.go]: added a TODO in https://github.com/uber/zanzibar/pull/57/files#diff-b99ff0337e82249d23a0d1420b13207fR60

File: examples/example-gateway/build/clients/baz/baz.go (hunk 5 above)
Comment (ChuntaoLu) [inline examples/example-gateway/build/clients/baz/baz.go]: fix in https://github.com/uber/zanzibar/pull/57/commits/a699af0c03603ed42332907c500f11f5ceef6daf

File: codegen/templates/tchannel_client.tmpl (hunk 6 above)
Comment (ChuntaoLu) [inline codegen/templates/tchannel_client.tmpl]: The test helper uses it, e.g. https://github.com/uber/zanzibar/pull/56/files#diff-c89f3c36e508e3ece8aa23fcd9e8e592R28

Comment (Raynos) [inline codegen/templates/tchannel_client.tmpl]: Ok can we move the interface into that file instead :)

File: codegen/templates/tchannel_client.tmpl (hunk 7 above)
Comment (ChuntaoLu) [inline codegen/templates/tchannel_client.tmpl]: fix in https://github.com/uber/zanzibar/pull/57/commits/e98eab15ecda1a3c5b792f1876335468c4fd678c

**Pull request merged**